File: src/com/android/settings/compatible/CompatibleListAdapter.java
package com.android.settings.compatible;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.CompoundButton;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.android.settings.R;
import com.android.settings.utils.LogTools;
import com.android.settings.utils.StringUtils;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import androidx.recyclerview.widget.RecyclerView;

public class CompatibleListAdapter extends RecyclerView.Adapter<CompatibleListAdapter.ViewHolder> {
    Context context;
    List<Map<String, Object>> list;

    String packageName;

    OnItemClickListener onItemClickListener;

    public CompatibleListAdapter(Context context, String packageName, List<Map<String, Object>> list,
            OnItemClickListener onItemClickListener) {
        this.context = context;
        this.list = list;
        this.packageName = packageName;
        this.onItemClickListener = onItemClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_compatible_list, parent, false);
        return new CompatibleListAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Map<String, Object> mp = list.get(position);

        String keyCode = StringUtils.ToString(mp.get("KEY_CODE"));
        holder.txtKey.setText(keyCode);
        String keyDescStr = CompUtils.parseEnChJson(context, StringUtils.ToString(mp.get("KEY_DESC")));
        holder.txtTitle.setText(keyDescStr);

        holder.rootView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // LogTools.i("onClick rootView " + position + " ,mp " + mp.toString());
                if (packageName == null || "".equals(packageName)) {

                } else {
                    onItemClickListener.onItemClick(position, "");
                }
            }
        });

        String noteStr = CompUtils.parseEnChJson(context, StringUtils.ToString(mp.get("NOTES")));
        holder.imgRemarks.setVisibility("".equals(noteStr) ? View.GONE : View.VISIBLE);
        holder.imgRemarks.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // LogTools.i("onClick imgRemarks " + position + " ,packageName " +
                // packageName);

                TipDialog tipDialog = new TipDialog(context, noteStr);
                if (!tipDialog.isShowing()) {
                    tipDialog.show();
                }
            }
        });

        holder.switchDown.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (packageName == null || "".equals(packageName)) {
                    if (b) {
                        holder.recyclerView.setVisibility(View.VISIBLE);
                    } else {
                        holder.recyclerView.setVisibility(View.GONE);
                    }
                }
            }
        });

        if (packageName == null || "".equals(packageName)) {
            holder.switchDown.setVisibility(View.VISIBLE);
            holder.imgRight.setVisibility(View.GONE);
            LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context, RecyclerView.VERTICAL, false);
            holder.recyclerView.setLayoutManager(linearLayoutManager);
            List<Map<String, Object>> listItem = CompatibleConfig.queryValueListData(context, keyCode);
            if (listItem == null) {
                listItem = new ArrayList<>();
            }

            List<Map<String, Object>> listTemp = new ArrayList<>();

            for (Map<String, Object> mmm : listItem) {
                AppData appData = CompUtils.getAppInfo(context, StringUtils.ToString(mmm.get("PACKAGE_NAME")));
                if (appData == null) {
                    if (StringUtils.ToString(mmm.get("FIELDS1")).contains("VNC_")) {
                        listTemp.add(mmm);
                    }
                } else {
                    listTemp.add(mmm);
                }
            }

            CompatibleItemAdapter compatibleItemAdapter = new CompatibleItemAdapter(context, listTemp,
                    list.get(position));
            holder.recyclerView.setAdapter(compatibleItemAdapter);
        } else {
            holder.recyclerView.setVisibility(View.GONE);
            holder.switchDown.setVisibility(View.INVISIBLE);
            holder.imgRight.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        LinearLayout rootView;
        TextView txtKey;
        TextView txtTitle;
        RecyclerView recyclerView;
        ImageView imgRemarks;
        ImageView imgRight;
        Switch switchDown;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            txtKey = itemView.findViewById(R.id.txtKey);
            txtTitle = itemView.findViewById(R.id.txtTitle);
            rootView = itemView.findViewById(R.id.rootView);
            imgRemarks = itemView.findViewById(R.id.imgRemarks);
            imgRight = itemView.findViewById(R.id.imgRight);
            switchDown = itemView.findViewById(R.id.switchDown);
            recyclerView = itemView.findViewById(R.id.recyclerView);
        }
    }
}

File: src/com/android/settings/compatible/CompatibleListAdapter.kt
package com.android.settings.compatible;

import android.content.Context
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Switch
import android.widget.TextView
import android.widget.PopupWindow
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.android.settings.R;
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import android.util.Log

class CompatibleListAdapter(
    private val context: Context,
    private val packageName: String,
    private val keyCode: String,
    private val activityName:String,
    private val onItemClickListener: OnItemClickListener
) : RecyclerView.Adapter<CompatibleListAdapter.ViewHolder>() {
    private var list = emptyList<CompatibleList>()
    private lateinit var popupWindow: PopupWindow;
    private val TAG = "CompatibleListAdapter"

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_compatible_list, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = list[position]
        val keyDescStr = CompUtils.parseEnChJson(context, item.keyDesc)
        holder.txtTitle.text = keyDescStr


        holder.recyclerView?.layoutManager = LinearLayoutManager(context);
  
        if (packageName != null && !"".equals(packageName)) {
            val adapter = CompatiblePageListAdapter(context, packageName, item)
            holder.recyclerView?.adapter = adapter;
            GlobalScope.launch(Dispatchers.IO) {
                var uniqueList = CompatibleConfig.queryValueListByKeyCodeAndPackageName(
                    context,
                    packageName,
                    item.keyCode
                );
                Log.w(TAG,"queryValueListByKeyCodeAndPackageName packageName "+packageName + ",item.keyCode : "+item.keyCode +",inputType: "+item.inputType)

                if(keyCode !=null && !"".equals(keyCode)){
                    //如果是从底部状态栏跳转过来
                    if(uniqueList == null || uniqueList.size  == 0){
                        uniqueList =  mutableListOf()
                    }
                    val containsItem = uniqueList.any {it.packageName == packageName && it.keyCode == keyCode && it.activityName == activityName}
                    if(!containsItem){
                        var compatibleValue: CompatibleValue = CompatibleValue();
                        compatibleValue.packageName = packageName
                        compatibleValue.keyCode = keyCode
                        compatibleValue.value = "false"
                        compatibleValue.activityName = activityName
                        uniqueList.add(compatibleValue);
                    }
                }else{
                    //从开始菜单跳转过来
                    if(uniqueList == null || uniqueList.size  == 0){
                        uniqueList =  mutableListOf()
                        var compatibleValue: CompatibleValue = CompatibleValue();
                        compatibleValue.packageName = packageName
                        compatibleValue.keyCode = item.keyCode
                        compatibleValue.value = ""
                        compatibleValue.activityName = activityName
                        uniqueList.add(compatibleValue);
                    }
                }
                
                
                withContext(Dispatchers.Main) {
                    adapter?.setData(uniqueList!!)
                }
            }
        } else {
            //from settings list 
            val adapter = CompatiblePkgListAdapter(context, packageName, item)
            holder.recyclerView?.adapter = adapter;

            GlobalScope.launch(Dispatchers.IO) {
                var listPkgs = CompatibleConfig.queryValueListByKeyCode(context, item.keyCode)
                var uniqueList = listPkgs?.distinctBy { it.packageName }

                withContext(Dispatchers.Main) {
                    adapter?.setData(uniqueList!!)
                }
            }
        }


        val noteStr = CompUtils.parseEnChJson(context, item.notes)
        holder.imgRemarks?.setOnHoverListener { v: View, event: MotionEvent ->
            when (event.action) {
                MotionEvent.ACTION_HOVER_ENTER -> {
                    showPopupWindow(context, v, noteStr)
                    true
                }

                MotionEvent.ACTION_HOVER_EXIT -> {
                    if (popupWindow != null) {
                        popupWindow.dismiss()
                    }
                    true
                }

                else -> false
            }
        }

        holder.switchDown.setOnCheckedChangeListener { _, isChecked ->
            holder.recyclerView.visibility = if (isChecked) View.VISIBLE else View.GONE
        }


    }

    fun setData(newData: List<CompatibleList>) {
        list = newData
        notifyDataSetChanged()
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int, keyCode: String)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val rootView: LinearLayout = itemView.findViewById<LinearLayout>(R.id.rootView)
            ?: throw IllegalArgumentException("ImagLinearLayouteView not found")
        val txtTitle: TextView = itemView.findViewById<TextView>(R.id.txtTitle)
            ?: throw IllegalArgumentException("TextView not found")
        val recyclerView: RecyclerView = itemView.findViewById<RecyclerView>(R.id.recyclerView)
            ?: throw IllegalArgumentException("RecyclerView not found")
        val imgRemarks: ImageView = itemView.findViewById<ImageView>(R.id.imgRemarks)
            ?: throw IllegalArgumentException("ImageView not found")
        val switchDown: Switch = itemView.findViewById<Switch>(R.id.switchDown)
            ?: throw IllegalArgumentException("Switch not found")
    }

    private fun showPopupWindow(context: Context, view: View, content: String) {
        val popupView = LayoutInflater.from(context).inflate(R.layout.popup_tip_view, null)
        popupWindow = PopupWindow(
            popupView,
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        popupView.findViewById<TextView>(R.id.txtContent)?.text = content;
        popupWindow.isOutsideTouchable = true
        popupWindow.showAsDropDown(view)
    }
}